compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4008,9 +4008,18 @@ public void visit(BLangBinaryExpr binaryExpr) {
 
         checkDecimalCompatibilityForBinaryArithmeticOverLiteralValues(binaryExpr);
 
+        BType nilLiftType = null;
+
+        if (isNullableBinaryExpr(binaryExpr)) {
+            BUnionType exprBType = (BUnionType) binaryExpr.expectedType;
+            nilLiftType = exprBType.getMemberTypes().iterator().next();
```

Comment:
```suggestion
            nilLiftType = ((BUnionType) binaryExpr.expectedType).getMemberTypes().iterator().next();
```

Reply:
Isn't this assuming the first type is always the non-nil type? This is not correct IMO. A union can have any number of members too.

Reply:
Yes, now the logic is changed. Now it works despite the number of types in the `union` type. And now this also works. 
`int|() or ()|int`

Reply:
The following still doesn't work

```ballerina
type T1 1|()|2;

type T2 1|2?;

public function main() {
    T1 a = 1;
    T2 b = 2;
    ()|int c = ();

    int? v = a % a; // error: operator '%' not defined for 'T1' and 'T1'
    v = a % b; // error: operator '%' not defined for 'T1' and 'T2'
    v = a % c; // error: operator '%' not defined for 'T1' and 'int?'
    v = b % b; // OK
    v = b % c; // error: operator '%' not defined for 'T2' and 'int?'
    v = c % c; // OK 
}
```